### PR TITLE
Adding OLMo

### PR DESCRIPTION
**Pull request by aflah02**
PR to support OLMo. Heavily borrows from https://github.com/Lightning-AI/litgpt/pull/927

~~Still WIP~~

Core code changes done. Working on training recipes

### CODE REVIEW

**Review by rasbt — COMMENTED** [review]
I know this is still WIP, but this looks good to me so far!

**Comment by Andrei-Aksionov** [comment]
Hello @aflah02 

Thanks for working on it.
Your PR seems to be at the finish line.
The only part that you need to carry from that PR is changes in the `tokenizer.py` [here](https://github.com/Lightning-AI/litgpt/pull/927/files#diff-c09fad817049a535518b7df27342baad0ad84a51e38d80964d91a77a7039d110R562-R565)

**Comment by aflah02** [comment]
> Hello @aflah02
> 
> Thanks for working on it. Your PR seems to be at the finish line. The only part that you need to carry from that PR is changes in the `tokenizer.py` [here](https://github.com/Lightning-AI/litgpt/pull/927/files#diff-9113bf8b641ca61e6ef212b777882d430eb7ddf3a0f5f4739e1af5d00ab12fe8R127-R129)

Makes sense! One follow-up question -- post training to export the lit checkpoint to a HF compatible OLMo class (https://huggingface.co/docs/transformers/en/model_doc/olmo#transformers.OlmoForCausalLM) don't we need custom export logic to assign this class to the exported model? As the reverse conversion (HF to lit checkpoint) piggybacks on the llama code but how will the other way around map the model to the right class?

**Comment by aflah02** [comment]
Ah nvm I just saw point 5 in this tutorial - https://github.com/Lightning-AI/litgpt/blob/main/tutorials/convert_lit_models.md#a-finetuning-and-conversion-tutorial

Is this the expected loading code post conversion - 

```
import torch
from transformers import AutoModel

state_dict = torch.load('FINETUNED_AND_CONVERTED_OLMO_MODEL_PATH')
model = AutoModel.from_pretrained("allenai/OLMo-1b-hf", state_dict=state_dict)
```

**Comment by Andrei-Aksionov** [comment]
We have scripts that convert weights from HF --> LitGPT and LitGPT --> HF.
Since we already have those scripts for LlaMa based models, it should work for Olmo too.
(As we found out that Olmo is a special case of LlaMa models. At least Olmo-hf. The original is quite different.)

Conversion is basically a renaming of layer names and, if needed, splitting/combining attention layer weights. This is needed because LitGPT uses a combined QKV matrix, while some models have separate Q, K and V matrices.
Oh, and sometimes we need to split MLP weights, because some models might combine them.
Overall, it's not a sophisticated process 🙂. 

You need to also port tests from the original PR (forgot to mention it, my bad). They check numerical equivalence after the conversion. If they pass - the conversion works without issues.

---

> but rather just loads directly if you point to the custom model name on HF

If I understood you correctly, the first step is to convert from Lit to HF format, then use HF tools for uploading weights to their hub.
I'm not that familiar with HF transformers, but I believe they provide an option to upload weights through CLI tools.
After weights are uploaded, you can just provide `repo_name/model_name` and it should work.

---

Bottom line: after weights are converted, they can be utilized by HF transformers like any other weights on their platform.
Whether you want to just load them into transformers, or to upload them to the Hub, or to quantize with any tool that expects HF weights format - it should work.

If it doesn't - please open an issue 😉.

File: README.md [hunk 1]
```diff
@@ -132,6 +132,7 @@ Every model is written from scratch to maximize performance and remove layers of
 | Mixtral MoE | 8x7B | Mistral AI | [Mistral AI 2023](https://mistral.ai/news/mixtral-of-experts/)                                                                     |
 | Mistral | 7B, 123B | Mistral AI | [Mistral AI 2023](https://mistral.ai/news/announcing-mistral-7b/)                                                                  |
 | Nous-Hermes | 7B, 13B, 70B | NousResearch | [Org page](https://huggingface.co/NousResearch)                                                                          |
+| OLMo | 1B, 7B | Allen Institute for AI (AI2) | [OLMo project page](https://allenai.org/olmo)    |
```
Comment (rasbt) [inline README.md:135]:
```suggestion
| OLMo | 1B, 7B | Allen Institute for AI (AI2) | [OLMo project page](https://arxiv.org/abs/2402.00838)    |
```

Let's use the OLMo paper here because it's hard to find any model information on the website.

File: tutorials/download_model_weights.md [hunk 2]
```diff
@@ -27,6 +27,7 @@ LitGPT supports a variety of LLM architectures with publicly available weights.
 | Mixtral MoE | 8x7B | Mistral AI | [Mistral AI 2023](https://mistral.ai/news/mixtral-of-experts/)                                                                     |
 | Mistral | 7B, 123B | Mistral AI | [Mistral AI 2023](https://mistral.ai/news/announcing-mistral-7b/)                                                                        |
 | Nous-Hermes | 7B, 13B, 70B | NousResearch | [Org page](https://huggingface.co/NousResearch)                                                                          |
+| OLMo | 1B, 7B | Allen Institute for AI (AI2) | [ OLMo project page](https://allenai.org/olmo)     |
```
Comment (rasbt) [inline tutorials/download_model_weights.md:30]:
```suggestion
| OLMo | 1B, 7B | Allen Institute for AI (AI2) | [ OLMo project page](https://arxiv.org/abs/2402.00838)     |
```

File: README.md (hunk 1 above)
Comment (aflah02) [reply README.md]: Gotcha
Also since the paper is accepted to ACL I'm pointing it to the paper uploaded on ACL Anthology - https://aclanthology.org/2024.acl-long.841/
I don't have any strong feelings eitherways but just thought this might be better

**Comment by aflah02** [comment]
@Andrei-Aksionov @rasbt 
When I copy over tests for checkpoint conversion I should use `copy_weights_llama` right? Since as noted [here](https://github.com/Lightning-AI/litgpt/pull/927#issuecomment-2462863923) the old olmo conversion code is just a special case of `copy_weights_llama`

**Comment by aflah02** [comment]
Also for now I've added both 7B and 7B Instruct into the tests but I guess since they're architecturally identical so we can remove one right?

**Comment by Andrei-Aksionov** [comment]
> When I copy over tests for checkpoint conversion I should use copy_weights_llama right?

Correct.

> Also for now I've added both 7B and 7B Instruct into the tests but I guess since they're architecturally identical so we can remove one right?

Also correct.
The difference should only in the training dataset and in the training process. The architecture should be the same.

----

Speaking of training.
I guess the next step is to add training recipes. Is it correct @rasbt?

**Comment by aflah02** [comment]
> > Speaking of training.
> > I guess the next step is to add training recipes. Is it correct @rasbt?
> 
> Historically we added the yaml config files but this would not be strictly necessary because it can take several hours to run and would be expensive for contributors. So please feel to skip. Someone else from the team can generate those.
> 
> But I suggest doing a quick test run with LoRA to make sure it seems to converge at least.

I'm looking at pretraining OLMo based models so I'll be happy to work on the yaml config files as well. I'll try to run some over the weekend/early next week and share the results here

> The other thing is I would also do a quick sanity check with `litgpt generate` and `litgpt chat`. Maybe with the prompt "What do Llamas eat?". You can post the responses here in the GitHub PR.
> 

I did run these but pasted them in the Old PR instead. You can find them here - https://github.com/Lightning-AI/litgpt/pull/927#issuecomment-2462456373

**Comment by aflah02** [comment]
I am thinking of training a TinyOLMo model same as TinyLLama. I copied over the config and made minor changes - 

```

# The name of the model to pretrain. Choose from names in ``litgpt.config``. Mutually exclusive with
# ``model_config``. (type: Optional[str], default: null)
model_name: OLMo-1b-hf

# A ``litgpt.Config`` object to define the model architecture. Mutually exclusive with
# ``model_config``. (type: Optional[Config], default: null)
model_config:

# Directory in which to save checkpoints and logs. If running in a Lightning Studio Job, look for it in
# /teamspace/jobs/<job-name>/share. (type: <class 'Path'>, default: out/pretrain)
out_dir: out/pretrain/tiny-olmo

# The precision to use for pretraining. Possible choices: "bf16-true", "bf16-mixed", "32-true". (type: Optional[str], default: null)
precision: bf16-mixed

# Optional path to a checkpoint directory to initialize the model from.
# Useful for continued pretraining. Mutually exclusive with ``resume``. (type: Optional[Path], default: null)
initial_checkpoint_dir:

# Path to a checkpoint directory to resume from in case training was interrupted, or ``True`` to resume
# from the latest checkpoint in ``out_dir``. An error will be raised if no checkpoint is found. Passing
# ``'auto'`` will resume from the latest checkpoint but not error if no checkpoint exists.
# (type: Union[bool, Literal["auto"], Path], default: False)
resume: false

# Data-related arguments. If not provided, the default is ``litgpt.data.TinyLlama``.
data: TinyLlama

# Training-related arguments. See ``litgpt.args.TrainArgs`` for details
train:

  # Number of optimizer steps between saving checkpoints (type: Optional[int], default: 1000)
  save_interval: 1000

  # Number of iterations between logging calls (type: int, default: 1)
  log_interval: 1

  # Number of samples between optimizer steps across data-parallel ranks (type: int, default: 512)
  global_batch_size: 512

  # Number of samples per data-parallel rank (type: int, default: 4)
  micro_batch_size: 4

  # Number of iterations with learning rate warmup active (type: int, default: 2000)
  lr_warmup_steps: 2000

  # Number of epochs to train on (type: Optional[int], default: null)
  epochs:

  # Total number of tokens to train on (type: Optional[int], default: 3000000000000)
  max_tokens: 3000000000000

  # Limits the number of optimizer steps to run. (type: Optional[int], default: null)
  max_steps:

  # Limits the length of samples. Off by default (type: Optional[int], default: null)
  max_seq_length: 2048

  # Whether to tie the embedding weights with the language modeling head weights. (type: Optional[bool], default: False)
  tie_embeddings:

  #   (type: Optional[float], default: 1.0)
  max_norm: 1.0

  #   (type: float, default: 4e-05)
  min_lr: 4.0e-05

# Evaluation-related arguments. See ``litgpt.args.EvalArgs`` for details
eval:

  # Number of optimizer steps between evaluation calls (type: int, default: 1000)
  interval: 1000

  # Number of tokens to generate (type: Optional[int], default: null)
  max_new_tokens:

  # Number of iterations (type: int, default: 100)
  max_iters: 100

  # Whether to evaluate on the validation set at the beginning of the training
  initial_validation: false

  # Whether to evaluate on the validation set at the end the training
  final_validation: false

# Optimizer-related arguments
optimizer:

  class_path: torch.optim.AdamW
  
  init_args:
    
    #   (type: float, default: 0.001)
    lr: 4e-4
    
    #   (type: float, default: 0.01)
    weight_decay: 0.1
    
    #   (type: tuple, default: (0.9,0.999))
    betas:
      - 0.9
      - 0.95

# How many devices/GPUs to use. Uses all GPUs by default. (type: Union[int, str], default: auto)
devices: auto

# How many nodes to use. (type: int, default: 1)
num_nodes: 1

# Optional path to the tokenizer dir that was used for preprocessing the dataset. Only some data
# module require this. (type: Optional[Path], default: null)
tokenizer_dir: checkpoints/allenai/OLMo-1b-hf

# The name of the logger to send metrics to. (type: Literal['wandb', 'tensorboard', 'csv'], default: tensorboard)
logger_name: wandb

# The random seed to use for reproducibility. (type: int, default: 42)
seed: 42
```

Does this work?

**Comment by aflah02** [comment]
Also should I continue this as a part of this PR? or should this be merged in first as the pretrain tests might take longer and I can do a few quick finetuning tests

**Comment by aflah02** [comment]
Also there seem to be quite a few other OLMo versions too. Should I add them in? See: https://huggingface.co/collections/allenai/olmo-suite-65aeaae8fe5b6b2122b46778

**Comment by aflah02** [comment]
One final thing I just realized. The names for the OLMo models seems very slightly different e.g. OLMo-7B-hf (On HF) v/s OLMo-7b-hf (in the PR), It seems HF is not case sensitive as both work but I guess keeping the same name as HF would be better? Just confirming as I copied the names from the old PR which have a different case from that on HF.

**Comment by rasbt** [comment]
> I'm looking at pretraining OLMo based models so I'll be happy to work on the yaml config files as well. I'll try to run some over the weekend/early next week and share the results here

Thanks! I'd say test that the yaml files work, but no worries about waiting until the run finishes. The reason is the numbers won't be comparable to those in the existing table because they are based on different hardware.


> I did run these but pasted them in the Old PR instead. You can find them here

They look good!

> Also should I continue this as a part of this PR? or should this be merged in first as the pretrain tests might take longer and I can do a few quick finetuning tests

Oh, you don't need to do a complete pretraining run, just running it briefly to make sure that there is no bug or so.

> Also there seem to be quite a few other OLMo versions too. Should I add them in? See: https://huggingface.co/collections/allenai/olmo-suite-65aeaae8fe5b6b2122b46778

Good suggestion, but this would add too much clutter in my opinion.

A user who is interested in those can already download them via

```
litgpt download allenai/OLMo-1B-0724-hf \
  --model_name OLMo-1B-hf
```

> One final thing I just realized. The names for the OLMo models seems very slightly different e.g. OLMo-7B-hf (On HF) v/s OLMo-7b-hf (in the PR

Good catch. I am just seeing that we've been inconsistent with this in the config file. I would just use uppercase B as in the llama models.

Thanks for all your attention to detail here!


(Btw I will be out of office after today, so in this case, please feel free to merge when it is ready @Andrei-Aksionov )

**Comment by Andrei-Aksionov** [comment]
@rasbt I don't think that I have a permission to merge a PR.
Definitely not to override your `change request`.

@aflah02 PR looks good. 
The remaining part is:
> One final thing I just realized. The names for the OLMo models seems very slightly different e.g. OLMo-7B-hf (On HF) v/s OLMo-7b-hf (in the PR), It seems HF is not case sensitive as both work but I guess keeping the same name as HF would be better? Just confirming as I copied the names from the old PR which have a different case from that on HF.

----
Training recipes can be done in a separate PR (if you want to do it).

**Comment by aflah02** [comment]
Hi @Andrei-Aksionov 
I'll make the change and also setup the pre-training run to start tomorrow. In the meanwhile, I did run a little finetuning experiments by copying over the config from another model - https://wandb.ai/aflah/finetune-OLMo-7b-hf/overview

**Comment by aflah02** [comment]
Hi @Andrei-Aksionov 
I was trying to create a config which mimics the setting from the OLMo paper (See Table 5 Page 17 [https://aclanthology.org/2024.acl-long.841.pdf](https://aclanthology.org/2024.acl-long.841.pdf)) however some args seem to be missing in the litgpt arg file (https://github.com/Lightning-AI/litgpt/blob/main/litgpt/args.py). 

I then looked at the tinyllama config for reference and I could find the initial learning rate under the optimizer header but I can't find any docs on what all is supported there like scheduler etc. and how can I provide the scheduler to use

**Comment by aflah02** [comment]
Also maybe a separate PR would be best for the pretraining since it would take some back and forth discussion so can you merge this one first?

**Comment by Andrei-Aksionov** [comment]
Hello @aflah02 

I'll merge the PR.
Thanks for pushing it over the line! 🎉 

-----

> I then looked at the tinyllama config for reference and I could find the initial learning rate under the optimizer header but I can't find any docs on what all is supported there like scheduler etc. and how can I provide the scheduler to use

LR scheduler is only one: linear followed by cosine

https://github.com/Lightning-AI/litgpt/blob/33eab00da13684a75b02eb9297fa4065d3f85e1e/litgpt/finetune/full.py#L383-L387

I think it's fine if Olmo will use this scheduler.

**Comment by aflah02** [comment]
Thanks! I didn't do much tbh since the code was pretty much all there already :)

For the scheduler what is the recommended route to use a custom scheduler? I want to use the plain linear scheduler for some futue OLMo runs as that is what the paper also used.

**Comment by Andrei-Aksionov** [comment]
> Thanks! I didn't do much tbh since the code was pretty much all there already :)

Description of image

🤓 

-----

> For the scheduler what is the recommended route to use a custom scheduler? I want to use the plain linear scheduler for some futue OLMo runs as that is what the paper also used.

I think there is no need for now to add the exact same scheduler as in the paper.
The codebase should be feature rich and up-to-date, but also an easy to read and not overly complicated (especially if there is no particular reason for it).

**Comment by aflah02** [comment]
> Description of image
> 🤓

Haha thanks!

> > For the scheduler what is the recommended route to use a custom scheduler? I want to use the plain linear scheduler for some futue OLMo runs as that is what the paper also used.
> 
> I think there is no need for now to add the exact same scheduler as in the paper. The codebase should be feature rich and up-to-date, but also an easy to read and not overly complicated (especially if there is no particular reason for it).

One reason I worked on integrating OLMo is that my colleagues and I are planning to run some pretraining experiments with the OLMo architecture. We want to follow the AI2 paper as closely as possible, including their choice of scheduler, to assess stability when using litgpt.  Hence, we would strongly prefer to use the linear scheduler they used over a 2-phase scheduler.

**Comment by Andrei-Aksionov** [comment]
I see.
Then you need to create something like [this](https://github.com/Lightning-AI/litgpt/blob/22528bf29c116fb70b210a0c574507d2c62f6619/litgpt/finetune/full.py#L158), but for schedulers.
It should parse the config and instantiate the corresponding class.
But, in order to preserve the current functionality, it should be able to parse a list of schedulers.

**Comment by aflah02** [comment]
> I see. Then you need to create something like [this](https://github.com/Lightning-AI/litgpt/blob/22528bf29c116fb70b210a0c574507d2c62f6619/litgpt/finetune/full.py#L158), but for schedulers. It should parse the config and instantiate the corresponding class. But, in order to preserve the current functionality, it should be able to parse a list of schedulers.

Thanks! I'll take a stab at this
